x-pack/plugins/security/public/session/session_timeout.tsx

```diff
@@ -24,57 +28,168 @@ const GRACE_PERIOD_MS = 5 * 1000;
 const WARNING_MS = 60 * 1000;
 
 export interface ISessionTimeout {
-  extend(): void;
+  init(): void;
+  extend(url: string): void;
 }
 
 export class SessionTimeout {
-  private warningTimeoutMilliseconds?: number;
-  private expirationTimeoutMilliseconds?: number;
+  private channel: BroadcastChannel<SessionInfo>;
+  private elector: ReturnType<typeof LeaderElection.create>;
+  private sessionInfo?: SessionInfo;
+  private updateTimer?: number;
+  private warningTimer?: number;
+  private expirationTimer?: number;
   private warningToast?: Toast;
 
   constructor(
-    private sessionTimeoutMilliseconds: number | null,
     private notifications: NotificationsSetup,
     private sessionExpired: ISessionExpired,
     private http: HttpSetup
-  ) {}
+  ) {
+    const name = `${http.basePath.get()}/session_timeout`;
+    this.channel = new BroadcastChannel(name, { webWorkerSupport: false });
+    this.elector = LeaderElection.create(this.channel, {
+      fallbackInterval: 2000,
+      responseTime: 1000,
+    });
+  }
+
+  init() {
+    if (this.isAnon()) {
+      this.elector.die();
+      return;
+    }
+
+    // subscribe to a broadcast channel for session timeout messages
+    // this allows us to synchronize the UX across tabs and avoid repetitive API calls
+    this.channel.onmessage = this.handleSessionInfo;
+    this.elector.awaitLeadership().then(() => {
+      this.updateTimeouts();
+    });
+
+    // Triggers an initial call to the endpoint to get session info;
+    // when that returns, it will set the timeout
+    return this.getSessionInfo();
+  }
 
-  extend() {
-    if (this.sessionTimeoutMilliseconds == null) {
+  extend(url: string) {
+    // ignore these URLs as they are requested by this module
+    if (this.isAnon() || url.endsWith('/session/info') || url.endsWith('/session/extend')) {
```

Comment:
I originally added those additional checks before I made use of the system API header, as it was causing an infinite loop of calls to the `/session/info` endpoint.

As you pointed out, the `/session/info` check is now unnecessary because of the existing check! The `/session/extends` check does prevent an additional API call when the user clicks the button to extend an idle session.